### PR TITLE
ENG-712: vars and secrets integration

### PR DESCRIPTION
**Pull request by efiShtain**
This is a draft PR
please let me know what do you think

- added generic secret store independent of connection
- added aws secret store
- added db secret store (default in dev mode)
- create var isSecret true - creates the secret in store
- get var isSecret true - return var from db
- reveal - return var from db + secret value from store
- delete - fetch all secrets, if isSecret also remove from secret store

added some tests, need to add more

need to check edge cases (creating multiple secret, partial success, what should we do? )

on delete we delete from from db and only then from secret store so secrets can't be referenced 
but we might have to cleanup in case we failed to delete a secret and we have dangling ones

### Create secret

```bash
ak var create <key> <val> -p <project_name> -e <env_name> --secret
```

### Get secret value
 ```bash
ak var get <key> -p <project_name> -e <env_name> --reveal
```

### Create from manifest
NOT SUPPORTED - We don't want to allow people setting secrets in the manifest

### CODE REVIEW

File: internal/backend/secrets/svc.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	"errors"
+	"fmt"
+
+	"go.autokitteh.dev/autokitteh/internal/backend/db"
+	"go.uber.org/zap"
+)
+
+type Secrets interface {
+	Get(ctx context.Context, key string) (map[string]string, error)
+	Set(ctx context.Context, key string, value map[string]string) error
```
Comment (itayd) [inline internal/backend/secrets/svc.go]: k->V string:string

File: internal/backend/secrets/svc.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		}
+	case SecretProviderDatabase:
+		if provider, err = newDatabaseSecrets(z, db); err != nil {
+			z.Panic("invalid provider configuration", zap.Error(err))
+			return nil, ErrSecretsInvalidProviderConfiguration
+		}
+	}
+
+	if provider == nil {
+		return nil, ErrSecretsInvalidProvider
```
Comment (itayd) [inline internal/backend/secrets/svc.go]: sdkerrors.NewInvalidArguments

File: internal/backend/secrets/svc.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+func New(cfg *Config, z *zap.Logger, db db.DB) (Secrets, error) {
+	var (
+		provider Secrets
+		err      error
+	)
+
+	switch cfg.Provider {
+	case SecretProviderAWSSecretManager:
+		if provider, err = newAWSSecrets(z, cfg); err != nil {
+			z.Panic("invalid provider configuration", zap.Error(err))
```
Comment (itayd) [inline internal/backend/secrets/svc.go]: kill

File: internal/backend/secrets/svc.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+			return nil, ErrSecretsInvalidProviderConfiguration
+		}
+	case SecretProviderDatabase:
+		if provider, err = newDatabaseSecrets(z, db); err != nil {
+			z.Panic("invalid provider configuration", zap.Error(err))
```
Comment (itayd) [inline internal/backend/secrets/svc.go]: kill

File: internal/backend/secrets/svc.go (hunk 3 above)
Comment (efiShtain) [reply internal/backend/secrets/svc.go]: done

File: internal/backend/secrets/svc.go (hunk 2 above)
Comment (efiShtain) [reply internal/backend/secrets/svc.go]: done

File: internal/backend/secrets/svc.go (hunk 1 above)
Comment (efiShtain) [reply internal/backend/secrets/svc.go]: done

File: internal/backend/secrets/svc.go (hunk 4 above)
Comment (efiShtain) [reply internal/backend/secrets/svc.go]: done

File: internal/backend/vars/vars.go [hunk 5]
```diff
@@ -2,34 +2,90 @@ package vars
...
+func varSecretKey(secret sdktypes.Var) string {
+	return fmt.Sprintf("%s/%s", secret.ScopeID().AsID().UUIDValue().String(), secret.Name())
 }
 
 func (v *vars) Set(ctx context.Context, vs ...sdktypes.Var) error {
+	for i, va := range vs {
+		if va.IsSecret() {
+			key := varSecretKey(va)
+			if err := v.secrets.Set(ctx, key, va.Value()); err != nil {
+				//TODO: handle delete secrets that were already created
```
Comment (itayd) [inline internal/backend/vars/vars.go]: ticket

File: internal/backend/vars/vars.go [hunk 6]
```diff
@@ -2,34 +2,90 @@ package vars
...
+			}
+		}
+	}
+
+	return err
 }
 
 func (v *vars) Get(ctx context.Context, sid sdktypes.VarScopeID, names ...sdktypes.Symbol) (sdktypes.Vars, error) {
-	return v.db.GetVars(ctx, false, sid, names)
+	return v.db.GetVars(ctx, sid, names)
```
Comment (itayd) [inline internal/backend/vars/vars.go:68]: this will not return the secrets keys right? i think we should include the keys here.

File: internal/backend/vars/vars.go [hunk 7]
```diff
@@ -2,34 +2,90 @@ package vars
...
 func (v *vars) Set(ctx context.Context, vs ...sdktypes.Var) error {
+	for i, va := range vs {
+		if va.IsSecret() {
+			key := varSecretKey(va)
+			if err := v.secrets.Set(ctx, key, va.Value()); err != nil {
+				//TODO: handle delete secrets that were already created
+				return err
+			}
+
+			vs[i] = sdktypes.NewVar(va.Name(), key, true).WithScopeID(va.ScopeID())
```
Comment (itayd) [inline internal/backend/vars/vars.go:38]: ahhh i see you store the secret key. is this really necessary though?

**Review by itayd — APPROVED** [review]
lgtm, though i got a q inside

File: tests/cli/vars.txt
Comment (daabr) [inline tests/cli/vars.txt:1]: Could you convert this to a system test instead of CLI?

File: internal/backend/db/db.go [hunk 8]
```diff
@@ -150,4 +150,9 @@ type DB interface {
 	GetSignal(ctx context.Context, signalID string) (scheme.Signal, error)
 	RemoveSignal(ctx context.Context, signalID string) error
 	ListSignalsWaitingOnConnection(ctx context.Context, connectionID sdktypes.ConnectionID) ([]scheme.Signal, error)
+
+	// -----------------------------------------------------------------------
+	SetSecret(ctx context.Context, key string, value string) error
+	GetSecret(ctx context.Context, key string) (string, error)
```
Comment (daabr) [inline internal/backend/db/db.go:156]: Question: do we also want to add an atomic update and/or upsert in a separate PR?

File: internal/backend/secrets/svc.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+type Secrets interface {
+	Get(ctx context.Context, key string) (string, error)
+	Set(ctx context.Context, key string, value string) error
+	Delete(ctx context.Context, key string) error
+}
+
+const (
+	SecretProviderDatabase         = "db"
+	SecretProviderAWSSecretManager = "awsSecretManager"
```
Comment (daabr) [inline internal/backend/secrets/svc.go:20]: Just my personal opinion, but I think "aws" is enough for configs, instead of "awsSecretManager".

File: internal/backend/secrets/svc.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+	SecretProviderVault            = "vault"
```
Comment (daabr) [inline internal/backend/secrets/svc.go:21]: I don't see the Vault implementation in this PR. Please add it, I'm relying on it.

**Review by daabr — APPROVED** [review]
Approved - but please see my comments. I don't mind if you implement them in a separate PR. Specifically not losing Vault support is very important for me.

File: internal/backend/vars/vars.go (hunk 7 above)
Comment (efiShtain) [reply internal/backend/vars/vars.go:38]: it makes it simpler to list secrets and vars without calling secrets stores

File: internal/backend/secrets/svc.go (hunk 9 above)
Comment (efiShtain) [reply internal/backend/secrets/svc.go:20]: they also support system parameter store for secrets, otherwise I agree with you
this makes sure we know what we use

File: internal/backend/secrets/svc.go (hunk 10 above)
Comment (efiShtain) [reply internal/backend/secrets/svc.go:21]: https://linear.app/autokitteh/issue/ENG-815/re-add-vault-as-a-secret-store

will do in another pr

File: internal/backend/db/db.go (hunk 8 above)
Comment (efiShtain) [reply internal/backend/db/db.go:156]: I change set in aws implementation to overwrite existing secrets
other than that it should be atomic as far as I see

File: tests/cli/vars.txt
Comment (efiShtain) [reply tests/cli/vars.txt:1]: https://linear.app/autokitteh/issue/ENG-816/convert-secrets-test-from-cli-test-to-system-tests

opened a ticket for that, since I'm not sure how to do it and I prefer to do it as another step

File: internal/backend/vars/vars.go (hunk 5 above)
Comment (efiShtain) [reply internal/backend/vars/vars.go]: https://linear.app/autokitteh/issue/ENG-817/handle-dangling-secrets

this is a new kind of task, we should have back office maintenance jobs to clean up dangling secrets
